reference/tools: add ticdc open protocol docs (#2284)
pingcap/docs: reference/tools/ticdc/open-protocol.md

---
title: TiCDC Open Protocol
category: reference
---

# TiCDC Open Protocol

> TiCDC Open Protocol is a row-level data change notification protocol that provides data sources for monitoring, caching, full-text indexing, analysis engines, and master-slave replication between different databases. TiCDC complies to TiCDC Open Protocol and replicates data changes of TiDB to third-party data media such as MQ (Message Queue).

Reviewer: @leoppro Please clarify what is "数据媒介". I can't find any useful info on that term.

> TiCDC Open Protocol is a row-level data change notification protocol that provides data sources for monitoring, caching, full-text indexing, analysis engines, and master-slave replication between different databases. TiCDC complies to TiCDC Open Protocol and replicates data changes of TiDB to third-party data media such as MQ (Message Queue).

Author: According to the author, it means "可以存数据的地方 比如 数据库（TiDB、MySQL）、消息队列（Kafak）、各类的数据分析引擎（es）等等"，which can be translated as `data medium`.

TiCDC Open Protocol uses Event as the basic unit to replicate data change events to the downstream. The Event is divided into three categories:

* Row Changed Event: Represents the data change in a row. When a row is changed, this Event is sent and contains information about the changed row.
* DDL Event: Represents the DDL change. This Event is sent after a DDL statement is successfully executed in the upstream. The DDL Event is broadcasted to every MQ Partition.
* Resolved Event: Represents a special time point before which the Event received is complete.

## Restrictions

* In most cases, the Row Changed Event of a version is sent only once, but in special situations such as node failure and network partition, the Row Changed Event of the same version might be sent multiple times.

> * Row Changed Events of the same version first sent on the same table are incremented in the order of timestamps (TS) in the Event stream.

Reviewer: `same version` or `each version`?

Author: each version.

> * Resolved Events are periodically broadcasted to each MQ Partition. The Resolved Event means that any Event with TS less than Resolved Event TS has been sent to the downstream.

Reviewer: I feel `less than` is weird. Is `smaller than` or `earlier than` better?
```suggestion
* Resolved Events are periodically broadcasted to each MQ Partition. The Resolved Event means that any Event with a TS less than Resolved Event TS has been sent to the downstream.
```

    | :---------- | :----- | :---------------------------------- |
    | TS          | Number |  The timestamp of the transaction that performs the DDL change.    |
    | Schema Name | String |  The schema name of the DDL change, which might be an empty string.  |
    | Table Name  | String |  The table name of the DDL change, which might be am empty string. |

+ **Value:**

    ```
    {
        "q":<DDL Query>,
        "t":<DDL Type>
    }
    ```

    | Parameter       | Type   | Description           |
    | :-------- | :----- | :------------ |
    | DDL Query | String | DDL Query SQL |
    | DDL Type  | String | The DDL type. See [Column and DDL Type Codes](/reference/tools/ticdc/column-ddl-type.md) for details.    |

### Resolved Event

+ **Key:**

    ```
    {
        "ts":<TS>,
        "t":3
    }
    ```

    | Parameter         | Type   | Description                                         |
    | :---------- | :----- | :------------------------------------------ |
    | TS          | Number | The Resolved timestamp. Any TS less than this Event has been sent. |

+ **Value:** None

## Examples of the Event stream output

> This section shows and displays the output logs of Events.

Reviewer: ```suggestion
This section shows and displays the output logs of the Event stream.
```

Suppose that you execute the following SQL statement in the upstream and the MQ Partition number is 2:

{{< copyable "sql" >}}

```sql
CREATE TABLE test.t1(id int primary key, val varchar(16));
```

> From the following Log 1 and Log 3, you can see that the DDL Event is broadcasted to all MQ Partitions.

Reviewer: ```suggestion
From the following Log 1 and Log 3, you can see that the DDL Event is broadcasted to all MQ Partitions, and that the Resolved Event is periodically broadcasted to each MQ Partition.
```

```
1. [partition=0] [key="{\"ts\":415508856908021766,\"scm\":\"test\",\"tbl\":\"t1\",\"t\":2}"] [value="{\"q\":\"CREATE TABLE test.t1(id int primary key, val varchar(16))\",\"t\":3}"]
2. [partition=0] [key="{\"ts\":415508856908021766,\"t\":3}"] [value=]
3. [partition=1] [key="{\"ts\":415508856908021766,\"scm\":\"test\",\"tbl\":\"t1\",\"t\":2}"] [value="{\"q\":\"CREATE TABLE test.t1(id int primary key, val varchar(16))\",\"t\":3}"]
4. [partition=1] [key="{\"ts\":415508856908021766,\"t\":3}"] [value=]
```

Execute the following SQL statements in the upstream:

{{< copyable "sql" >}}

```sql
BEGIN;
INSERT INTO test.t1(id, val) VALUES (1, 'aa');
INSERT INTO test.t1(id, val) VALUES (2, 'aa');
UPDATE test.t1 SET val = 'bb' WHERE id = 2;
INSERT INTO test.t1(id, val) VALUES (3, 'cc');
COMMIT;
```

+ From the following Log 5 and Log 6, you can see that Row Changed Events on the same table might be sent to different partitions based on the primary key, but changes to the same row are sent to the same partition so that the downstream can easily process the Event concurrently.

> + From Log 6, multiple changes to the same row in a transaction only send one Row Changed Event.

Reviewer: ```suggestion
+ From Log 6, multiple changes to the same row in a transaction are only sent in one Row Changed Event.
```

> + Log 8 is a repeated event of Log 7. Row Changed Event might be repeated, but each version of a first Event is sent orderly.

Reviewer: @leoppro Please confirm whether this is what you mean.
```suggestion
+ Log 8 is a repeated event of Log 7. Row Changed Event might be repeated, but the first Event of each version is sent orderly.
```
